### PR TITLE
[#2357] improvement(trino): Fix varchar type mapping between Gravitino JDBC catalog and Trino

### PR DESCRIPTION
**Pull request by yuqi1129**
### What changes were proposed in this pull request?

Modify the type mapping for VARCHAR and CHAR to match Gravitino and Trino.

### Why are the changes needed?

please see issue #2356

Fix: #2035, #2357

### Does this PR introduce _any_ user-facing change?

N/A. 

### How was this patch tested?

New UTs.

### CODE REVIEW

**Comment by yuqi1129** [comment]
@diqiu50 
Could you please help me take a look? Thank you.

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/jdbc/mysql/MySQLDataTypeTransformer.java [hunk 1]
```diff
@@ -11,27 +11,62 @@
...
 import com.datastrato.gravitino.trino.connector.util.GeneralDataTypeTransformer;
 import io.trino.spi.TrinoException;
+import io.trino.spi.type.CharType;
 
 /** Type transformer between MySQL and Trino */
 public class MySQLDataTypeTransformer extends GeneralDataTypeTransformer {
   private static final int MYSQL_CHAR_LENGTH_LIMIT = 255;
+  // 65535 / 4 = 16383.
+  private static final int MYSQL_VARCHAR_LENGTH_LIMIT = 16383;
+
```
Comment (diqiu50) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/jdbc/mysql/MySQLDataTypeTransformer.java:21]: Can you give the reference of this limitation?

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/jdbc/mysql/MySQLDataTypeTransformer.java [hunk 2]
```diff
@@ -11,27 +11,62 @@
...
+    return super.getTrinoType(type);
+  }
 
   @Override
   public Type getGravitinoType(io.trino.spi.type.Type type) {
-    Type gravitinoType = super.getGravitinoType(type);
-    if (gravitinoType.name() == Name.VARCHAR) {
-      if (((Types.VarCharType) gravitinoType).length() > MYSQL_CHAR_LENGTH_LIMIT) {
+    Class<? extends io.trino.spi.type.Type> typeClass = type.getClass();
+    if (typeClass == io.trino.spi.type.CharType.class) {
```
Comment (diqiu50) [inline trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/jdbc/mysql/MySQLDataTypeTransformer.java:34]: Why don't we use the `instanceof` operator?

**Review by diqiu50 — COMMENTED** [review]
I suggest adding integration test cases for char/varchar and testing boundary conditions.

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/jdbc/mysql/MySQLDataTypeTransformer.java (hunk 1 above)
Comment (yuqi1129) [reply trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/jdbc/mysql/MySQLDataTypeTransformer.java:21]: OK

File: trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/jdbc/mysql/MySQLDataTypeTransformer.java (hunk 2 above)
Comment (yuqi1129) [reply trino-connector/src/main/java/com/datastrato/gravitino/trino/connector/catalog/jdbc/mysql/MySQLDataTypeTransformer.java:34]: I don't have any preference about `instanceof` and `getClass()`, as we will judge the type condition twice, so I think the the temporary variant and use the format `==` may be simple.

**Comment by yuqi1129** [comment]
@diqiu50 
Could you please help me take another look?

File: integration-test/src/test/resources/trino-ci-testset/testsets/jdbc-postgresql/00007_varchar.txt [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+)
+COMMENT ''"
+
+"CREATE TABLE ""test.gt_postgresql"".varchar_db1.tb02 (
+   id integer,
+   name char(65536)
+)
+COMMENT ''"
+
+<QUERY_FAILED> Unknown type 'char(65537)' for column 'name'
```
Comment (diqiu50) [inline integration-test/src/test/resources/trino-ci-testset/testsets/jdbc-postgresql/00007_varchar.txt:19]: I think the error message should not be understandable by the user

File: integration-test/src/test/resources/trino-ci-testset/testsets/jdbc-postgresql/00007_varchar.txt [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+CREATE TABLE
+
+"CREATE TABLE ""test.gt_postgresql"".varchar_db1.tb05 (
+   id integer,
+   name varchar(10485760)
+)
+COMMENT ''"
+
+<QUERY_FAILED> Failed to create table.
```
Comment (diqiu50) [inline integration-test/src/test/resources/trino-ci-testset/testsets/jdbc-postgresql/00007_varchar.txt:37]: Why do we show more detailed messages for the error

File: integration-test/src/test/resources/trino-ci-testset/testsets/jdbc-postgresql/00007_varchar.txt [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
```
Comment (diqiu50) [inline integration-test/src/test/resources/trino-ci-testset/testsets/jdbc-postgresql/00007_varchar.txt:38]: Missing  line 25 command output, Does it still work with subsequent commands?
```
SHOW CREATE TABLE "test.gt_postgresql".varchar_db1.tb06;
```

File: integration-test/src/test/resources/trino-ci-testset/testsets/jdbc-mysql/00007_varchar.sql [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+SHOW CREATE TABLE "test.gt_mysql".varchar_db1.tb02;
+
+CREATE TABLE tb03 (id int, name char(256));
+
+CREATE TABLE tb04 (id int, name varchar(250));
+
+SHOW CREATE TABLE "test.gt_mysql".varchar_db1.tb04;
+
+CREATE TABLE tb05 (id int, name varchar(256));
+
```
Comment (diqiu50) [inline integration-test/src/test/resources/trino-ci-testset/testsets/jdbc-mysql/00007_varchar.sql:20]: Missing the test cases:
```
CREATE TABLE tb01 (id int, name char);
CREATE TABLE tb01 (id int, name varchar);
```

File: integration-test/src/test/resources/trino-ci-testset/testsets/jdbc-postgresql/00007_varchar.txt (hunk 3 above)
Comment (yuqi1129) [reply integration-test/src/test/resources/trino-ci-testset/testsets/jdbc-postgresql/00007_varchar.txt:19]: The error message is from Trino itself, not from the Gravitino connector.

File: integration-test/src/test/resources/trino-ci-testset/testsets/jdbc-postgresql/00007_varchar.txt (hunk 5 above)
Comment (yuqi1129) [reply integration-test/src/test/resources/trino-ci-testset/testsets/jdbc-postgresql/00007_varchar.txt:38]: OK

**Review by diqiu50 — APPROVED** [review]
LGTM
